src/connector/src/connector_common/common.rs

```diff
@@ -901,23 +935,78 @@ impl NatsCommon {
         Ok(client)
     }
 
+    /// Build a NATS client, attempting to reuse an existing cached client if available.
+    /// The client is cached using `Weak` references, so it will be cleaned up when all
+    /// strong references are dropped.
+    pub(crate) async fn build_client(&self) -> ConnectorResult<Arc<async_nats::Client>> {
+        let connection_props = self.connection_props();
+        let mut client: Option<Arc<async_nats::Client>> = None;
+
+        SHARED_NATS_CLIENT
+            .entry_by_ref(&connection_props)
+            .and_try_compute_with::<_, _, crate::error::ConnectorError>(|maybe_entry| async {
+                if let Some(entry) = maybe_entry {
+                    let entry_value = entry.into_value();
+                    if let Some(existing_client) = entry_value.upgrade() {
+                        // Check if the connection is still healthy
+                        if existing_client.connection_state()
+                            == async_nats::connection::State::Connected
+                        {
+                            tracing::info!("reuse existing nats client for {}", self.server_url);
+                            client = Some(existing_client);
+                            return Ok(Op::Nop);
+                        }
+                        // Connection is not healthy, create a new one
+                        tracing::info!(
+                            "existing nats client for {} is not connected, creating new one",
+                            self.server_url
+                        );
+                    }
+                }
+                tracing::info!("build new nats client for {}", self.server_url);
+                let new_client = Arc::new(self.build_client_inner().await?);
+                client = Some(new_client.clone());
+                Ok(Op::Put(Arc::downgrade(&new_client)))
```

Comment:
I think we can improve the code structure here.

Reply:
Pushed https://github.com/risingwavelabs/risingwave/pull/23963/commits/561a89315b0eeca370b07756c86c18ae2fd35d64